### PR TITLE
Implement `recursive` option in `fs.readdir`

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

This implements support for `readdir(path, {recursive: true})`

```js
import {readdir} from 'fs/promises';
const results = await readdir(__dirname, { recursive: true });
console.log(results) // ["a.js", "b/c.js" ...
```

### How did you verify your code works?

There are tests.

### CODE REVIEW

File: src/bun.js/node/node_fs.zig [hunk 1]
```diff
@@ -317,6 +319,349 @@ pub const AsyncCpTask = struct {
...
 
+pub const AsyncReaddirRecursiveTask = struct {
+    promise: JSC.JSPromise.Strong,
+    args: Arguments.Readdir,
+    globalObject: *JSC.JSGlobalObject,
+    task: JSC.WorkPoolTask = .{ .callback = &workPoolCallback },
+    ref: bun.Async.KeepAlive = .{},
+    tracker: JSC.AsyncTaskTracker,
+
+    // It's not 100% clear this one is necessary
```
Comment (ishank-s) [inline src/bun.js/node/node_fs.zig:330]: hi @Jarred-Sumner  , wanted to know your opinion on comments like these?
are they cool for all code bases or you add these cause you're the owner. 
Thanks for your great work!

File: src/bun.zig [hunk 2]
```diff
@@ -2052,3 +2052,7 @@ pub fn exitThread() noreturn {
         @panic("Unsupported platform");
     }
 }
+
+pub fn outOfMemory() noreturn {
```
Comment (paperclover) [inline src/bun.zig:2056]: thank you for using this pattern.

File: src/bun.js/node/node_fs.zig [hunk 3]
```diff
@@ -317,6 +319,349 @@ pub const AsyncCpTask = struct {
...
+        if (result.items.len > 0) {
+            const Field = comptime switch (ResultType) {
+                bun.String => .files,
+                Dirent => .with_file_types,
+                Buffer => .buffers,
+                else => unreachable,
+            };
+            var list = bun.default_allocator.create(ResultListEntry) catch bun.outOfMemory();
+            errdefer {
+                bun.default_allocator.destroy(list);
```
Comment (paperclover) [inline src/bun.js/node/node_fs.zig:525]: nitpick: is this code ever run?

> Zig has `errdefer`, which evaluates the deferred expression on block exit path if and only if the function returned with an error from the block.

(there is no `return error` case)

File: src/bun.js/node/node_fs.zig [hunk 4]
```diff
@@ -4208,19 +4542,335 @@ pub const NodeFS = struct {
...
+                    return .{
+                        .err = err,
+                    };
+                },
+                .result => .{ .result = @unionInit(Return.Readdir, file_type, entries.items) },
+            };
+        }
+
+        if (comptime recursive) {
+            @panic("This code path should never be reached. It should only go through readdirWithEntriesRecursiveAsync.");
```
Comment (paperclover) [inline src/bun.js/node/node_fs.zig:4856]: can this be `@compileError`?

**Review by paperclover — APPROVED** [review]
no meaningful comments, this looks good

File: src/bun.js/node/node_fs.zig (hunk 4 above)
Comment (Jarred-Sumner) [reply src/bun.js/node/node_fs.zig:4856]: I tried. But no, unfortunately

**Comment by NReilingh** [comment]
Doesn't this forget to update the type definitions? My LSP is complaining when I pass `{ recursive: true }` on 1.0.15. I took a look at https://github.com/oven-sh/bun/blob/main/packages/bun-types/fs.d.ts and expect to see the `recursive` option alongside `withFileTypes`; only `withFileTypes` is present on the `readdir`/`readdirSync` calls.

**Pull request merged**